weldx/transformations.py

```diff
@@ -379,45 +518,449 @@ def _calculate_orthogonal_axis(a_0, a_1):
+        self._add_coordinate_system_node(root_coordinate_system_name)
+
+    def _add_coordinate_system_node(self, coordinate_system_name):
+        self._check_new_coordinate_system_name(coordinate_system_name)
+        self._graph.add_node(coordinate_system_name, data=[])
+
+    def _add_edges(
+        self, node_from: Hashable, node_to: Hashable, lcs: LocalCoordinateSystem
+    ):
+        """
+        Add an edge to the internal graph.
+
+        :param node_from: Start node of the edge
+        :param node_to: End node of the edge
+        :param lcs: Local coordinate system
+        :return: ---
+        """
+        self._graph.add_edge(node_from, node_to, lcs=lcs)
+        self._graph.add_edge(node_to, node_from, lcs=lcs.invert())
+
+    def _check_coordinate_system_exists(self, coordinate_system_name: Hashable):
+        """
+        Raise an exception if the specified coordinate system does not exist.
+
+        :param coordinate_system_name: Name of the coordinate system, that should be
+        checked.
+        :return: ---
+        """
+        if not self.has_coordinate_system(coordinate_system_name):
+            raise Exception(
+                "There is no coordinate system with name " + str(coordinate_system_name)
+            )
+
+    def _check_new_coordinate_system_name(self, coordinate_system_name: Hashable):
+        """
+        Raise an exception if the new coordinate systems' name is invalid.
+
+        :param coordinate_system_name: Name of the new coordinate system, that should be
+        checked.
+        :return: ---
+        """
+        if not isinstance(coordinate_system_name, cl.Hashable):
+            raise Exception("The coordinate system name must be a hashable type.")
+        if self.has_coordinate_system(coordinate_system_name):
+            raise Exception(
+                "There already is a coordinate system with name "
+                + str(coordinate_system_name)
+            )
+
+    def add_coordinate_system(
+        self,
+        coordinate_system_name: Hashable,
+        reference_system_name: Hashable,
+        local_coordinate_system: LocalCoordinateSystem,
+    ):
+        """
+        Add a coordinate system to the coordinate system manager.
+
+        :param coordinate_system_name: Name of the new coordinate system. This can be
+        any hashable type, but it is recommended to use strings.
+        :param reference_system_name: Name of the parent system. This must have been
+        already added.
+        :param local_coordinate_system: An instance of
+        weldx.transformations.LocalCoordinateSystem that describes how the new
+        coordinate system is oriented in its parent system.
+        :return: ---
+        """
+        if not isinstance(local_coordinate_system, LocalCoordinateSystem):
+            raise Exception(
+                "'local_coordinate_system' must be an instance of "
+                + "weldx.transformations.LocalCoordinateSystem"
+            )
```

Comment:
This (and similar cases) should raise [`TypeError`](https://docs.python.org/3.7/library/exceptions.html#TypeError).